pennylane/tape/unwrap.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a context manager for unwrapping tapes
+"""
+import pennylane as qml
+
+
+class UnwrapTape:
+    """A context manager that unwraps a tape with tensor-like parameters
+    to NumPy arrays.
+
+    Args:
+        tape (.QuantumTape): the quantum tape to unwrap
+
+    Returns:
+
+        .QuantumTape: the unwrapped quantum tape
+
+    **Example**
+
+    >>> with tf.GradientTape():
+    ...     with qml.tape.QuantumTape() as tape:
+    ...         qml.RX(tf.Variable(0.1), wires=0)
+    ...         qml.RY(tf.constant(0.2), wires=0)
+    ...         qml.RZ(tf.Variable(0.3), wires=0)
+    ...     with UnwrapTape(tape) as unwrapped_tape:
+    ...         print("Trainable params:", unwrapped_tape.trainable_params)
+    ...         print("Unwrapped params:", unwrapped_tape.get_parameters())
+    Trainable params: {0, 2}
+    Unwrapped params: [0.1, 0.3]
+    >>> print("Original parameters:", tape.get_parameters())
+    Original parameters: [<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=0.1>,
+      <tf.Variable 'Variable:0' shape=() dtype=float32, numpy=0.3>]
+    """
+
+    def __init__(self, tape):
+        self.tape = tape
+        self._original_params = None
+        self._unwrapped_params = None
+
+    def __enter__(self):
+        self._original_params = self.tape.get_parameters(trainable_only=False)
+        self.tape.trainable_params = qml.math.get_trainable_indices(self._original_params)

```

Comment:
This is not so urgent, but I often find it confusing that `trainable_params` are really the trainable indices. One day we may want to change the name!

Reply:
So this step is just filling this attribute, and does not have to be reset (because the information does not depend on wrapping), right? Would all applications of the new context need this attribute, so you just always set it here?

Reply:
> This is not so urgent, but I often find it confusing that trainable_params are really the trainable indices. One day we may want to change the name!

This PR made me think the same thing 🙂 

> So this step is just filling this attribute, and does not have to be reset (because the information does not depend on wrapping), right? Would all applications of the new context need this attribute, so you just always set it here?

This is a good point. Currently, all cases where we need to unwrap the tape, we _also_ need to get the trainable indices. So I do it here to save time. But I suppose it doesn't _have_ to be done here? Should I move it?

Reply:
Don't worry to leave it, maybe with a comment of why it is here?